woocommerce/woocommerce-ios
# Remove GhostableViewController Conformances for Crash Investigation

## Description


This attempts a fix to a top crash. See discussion on peaMlT-qR-p2. 

The root cause of the crash appears to be modifying the sublayers array of a CALayer while are iterating over it. An investigation revealed that the code causing the crash could be in `UIView+Ghost.swift` and `UIView+InnerGhost.swift`. Rather than attempting a fix directly which would require a fix in the WordPressUI pod, in this PR those types are used less often by removing conformances to `GhostableViewController`.

Of 13 total `GhostableViewController` conformances in the app 7 are removed here.

If the theory about the cause of the crash is correct the volume of "Collection CALayerArray: 0x2823ca370 was mutated while being enumerated crashes" should go down once this change is released.

Each `GhostableViewController` conformance removal was done on a view controller that also uses a `UIRefreshControl` for pull to refresh support. So while this change will remove the placeholder UI that `GhostableViewController` provides, the user has access to pull to refresh which if performed will reveal a loading indicator.

If this change can be merged and the crash goes down as predicted this PR can be reverted and a direct crash fix can be attempted later.

## Steps to reproduce
The crash so far has not been able to be reproduced.

## Testing information

Using the network link conditioner set to Very Bad Network allows this testing to be done more easily.

For any of the screens that no longer have the placeholder UI provided by `GhostableViewController`, tap into the screen and observe the placeholder is no longer shown.

During that time, swipe down on the screen to pull to refresh. Observe that a loading indicator appears.

A specific screen to try this on is the product variations screen: 

- tap into a variable product
- tap Variations
- observe no place holder UI is shown while the variations load, but that swiping down reveals a pull to refresh indicator.

To see the the placeholder UI try the product variations attributes screen which still conforms to `GhostableViewController`. 

- tap into a variable product
- tap Variations Attributes
- tap Add New Attribute
- observe that placeholder UI is shown briefly.

## Screenshots


For before and after images when adding attributes for a product variation see the comment "iOS CALayerArray Crash Investigation and Next Steps" on peaMlT-qR-p2


---
